Sources/TensorFlow/Optimizer.swift

```diff
@@ -38,16 +37,16 @@ fileprivate extension Tensor where Scalar: Numeric {
 ///
 /// Reference: ["Adam - A Method for Stochastic Optimization"](
 /// https://arxiv.org/abs/1412.6980v8)
-public class Adam<Model: Layer>: Optimizer
-    where Model.AllDifferentiableVariables == Model.TangentVector {
+public class Adam<Model: Differentiable>: Optimizer
+    where Model.TangentVector: VectorProtocol & PointwiseMultiplicative & ElementaryFunctions,
+          Model.TangentVector.VectorSpaceScalar == Float,
+          Model.AllDifferentiableVariables == Model.TangentVector {
```

Comment:
This constraint is no longer necessary.

Reply:
It does not compile without it:
```
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:82:23: error: property 'allDifferentiableVariables' requires the types 'Model' and 'Model.AllDifferentiableVariables' be equivalent
        update(&model.allDifferentiableVariables, along: direction)
                      ^
Swift.Differentiable:1:11: note: where 'Self' = 'Model', 'Self.AllDifferentiableVariables' = 'Model.AllDifferentiableVariables'
extension Differentiable where Self == Self.AllDifferentiableVariables {
          ^
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:96:57: error: binary operator '*' cannot be applied to operands of type 'Model.AllDifferentiableVariables' and 'Float'
        firstMoments = firstMoments * beta1 + direction * (1 - beta1)
                                              ~~~~~~~~~ ^ ~~~~~~~~~~~
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:96:57: note: expected an argument list of type '(Float, Float)'
        firstMoments = firstMoments * beta1 + direction * (1 - beta1)
                                                        ^
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:98:36: error: binary operator '.*' cannot be applied to two 'Model.AllDifferentiableVariables' operands
        secondMoments += direction .* direction * (1 - beta2)
                         ~~~~~~~~~ ^  ~~~~~~~~~
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:98:36: note: expected an argument list of type '(Tensor<Scalar>, Tensor<Scalar>)'
        secondMoments += direction .* direction * (1 - beta2)
                                   ^
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:101:15: error: binary operator '-=' cannot be applied to operands of type 'Model.AllDifferentiableVariables' and 'Model.TangentVector'
        model -= stepSize * firstMoments .* denominator.reciprocal
        ~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/eaplatanios/Development/GitHub/swift-apis/Sources/TensorFlow/Optimizer.swift:101:15: note: expected an argument list of type '(inout Self, Self)'
        model -= stepSize * firstMoments .* denominator.reciprocal
              ^
[1/2] Compiling TensorFlow Optimizer.swift
```

Reply:
Instead of `model -=`, `move(along:)` should be used.